sycl/source/detail/reduction.cpp

```diff
@@ -95,16 +123,10 @@ reduGetMaxWGSize(std::shared_ptr<sycl::detail::queue_impl> Queue,
   // the local memory assigned to one work-group by code in another work-group.
   // It seems the only good solution for this work-group detection problem is
   // kernel precompilation and querying the kernel properties.
-  if (WGSize >= 4) {
+  if (WGSize >= 4 && WGSizePerMem < MaxWGSize * 2) {
     // Let's return a twice smaller number, but... do that only if the kernel
-    // is limited by memory, or the kernel uses opencl:cpu backend, which
-    // surprisingly uses lots of resources to run the kernels with reductions
-    // and often causes CL_OUT_OF_RESOURCES error even when reduction
-    // does not use local accessors.
-    if (WGSizePerMem < MaxWGSize * 2 ||
-        (Queue->get_device().is_cpu() &&
-         Queue->get_device().get_platform().get_backend() == backend::opencl))
-      WGSize /= 2;
+    // is limited by memory.
+    WGSize /= 2;
```

Comment:
This division is over-insurance and is done just for extra safety used to unexplained problems on opencl/CPU.
Do you think it is needed for other devices/platforms?

Also, is is code reachable at all, taking into account all those early returns above (return for CPU, return for GPU and return for ACC above)?

Reply:
Of the 3, CPU is the only one to be guaranteed to early exit, as it either picks 16 or the configuration setting, which is also why I chose to remove the RHS of the disjunction here as it checked for CPU which would never happen with this. GPU and accelerator will only return early if their configurations are set.

Reply:
Right, I see, this code  is reachable for GPU/ACC.

My question then is: why do we reduce twice the WGSize for them? The original work-around is very odd and was supposed to work only for opencl/cpu, which had some bug causing low-resources error when usage of local-memory was close to max-wg-size.
I.e. if some GPU has 256-work-items, and local memory is enough to use them, then it is ok to use them.

I remember that my experiments on GPU showed that the bigger work-group sizes gave better perf results.

Reply:
Did we add consistent testing for it or was it too hardware-specific?

Reply:
> "I simply removed the CPU case as it was now unreachable"

The "WGSize /= 2" was called only for CPU before the patch. The current patch makes it reachable for GPU as well, which does not seem right.